Fix legacy salmon and add experiment class for salmon

dyokelson opened the pull request:
## Description

Removing dependence on the Fujitsu compiler.

This PR works on ruby with legacy salmon, but needs testing on Fugaku. I had to remove `-stdin` from the `application.py`, and I'm concerned this is required by Fugaku but not other systems. Here are the steps to reproduce with this branch:
```
cd benchpark
. setup-env.sh
benchpark setup salmon/openmp RCCS-Fugaku-Fujitsu-A64FX-TofuD workspace
```

Then follow the steps in output

Stephanie TODO:
* (DONE) Create `experiment.py` and verify generated `ramble.yaml` with hand written one in this PR
* (DONE) Replace dry run with `dynamic cts ruby`
* (DONE) Add dry run for fugaku (new way)

With the new way, these changes are included in this PR:
* Update spack version
* Remove babelstream package in benchpark in favor of using more recently updated package in spack
* Adds `-fallow-argument-mismatch`  for gcc10 and later
* Creates benchpark salmon-tddft package and related patches with openmp variant

Generated `ramble.yaml`. It does not generate a section for `rt` workload.
```
ramble:
  applications:
    salmon-tddft:
      workloads:
        gs:
          experiments:
            salmon-tddft_gs_single_node_openmp_{n_nodes}:
              exclude: {}
              matrix: []
              variables:
                arch: OpenMP
                exercise: exercise_01_C2H2_gs
                experiment_setup: ''
                inp_file: C2H2_gs.inp
                n_nodes:
                - '1'
                n_ranks: '{processes_per_node} * {n_nodes}'
                omp_num_threads:
                - '12'
                processes_per_node:
                - '4'
              variants:
                package_manager: spack
              zips: {}
  config:
    deprecated: true
    spack_flags:
      concretize: -U -f
      install: --add --keep-stage
  include:
  - ./configs
  modifiers:
  - name: allocation
  - name: exit-code
  software:
    environments:
      salmon-tddft:
        packages:
        - salmon-tddft
    packages:
      salmon-tddft:
        compiler: default-compiler
        pkg_spec: salmon-tddft@master+openmp
```

Dependencies: None

Fixes issue(s): Part of #460 

## Type of Change

- { } Adding a system, benchmark, or experiment
- {x} Modifying an existing system, benchmark, or experiment
- { } Documentation update
- {x} Build/CI update
- { } Benchpark core functionality

## Checklist:

If adding/modifying a benchpark:
- {x} Add a new `application.py` and (maybe) `package.py` under a new directory
      for this benchmark
- {x} Configure an experiment
- {x} Benchmark appears in Benchmarks and Experiments table in docs catalogue
      section

If adding/modifying a experiment:
- {x} Extend `experiment.py` under existing directory for specific benchmark
- {x} Define a single node and multi-node experiments

If adding/modifying core functionality:
- {x} Update docs
- {x} Update `.github/workflows` and `.gitlab/ci` unit tests (if needed)

dyokelson committed:
change to default-compiler, try out on dane tomorrow

dyokelson committed:
Add CI on x86 and default-compiler to cmake

dyokelson committed:
remove .tpl file

slabasan committed:
running on ruby

use a consistent compiler between ruby and fugaku

pearce8 commented on .github/workflows/run.yml:
@slabasan does it work on dynamic systems?  Can you add a dynamic system dry run?

slabasan replied:
Added it now along with the `application.py`. Need to confirm it's correct through github CI.

slabasan committed:
add salmon application.py

slabasan committed:
add dynamic dry runs on ruby and fugaku

slabasan committed:
rename salmon directory, class and format

slabasan committed:
salmon: add openmp variant

slabasan committed:
add local salmon package and patches

remove variant from application.py file

slabasan committed:
fix spack package name for salmon

slabasan committed:
bump spack to nov release

slabasan committed:
use spack babelstream, package recently redone

slabasan committed:
salmon package.py and experiment.py fixes

- openmp is not a package in spack, so cannot depend_on()
- add fallow-argument-mismatch flag for gcc10 and later
